### ISSUE TITLE
Test failures for GCC 8

### ISSUE THREAD

**Issue by rouson** [sourceryinstitute/OpenCoarrays#447]
[issue stat img]: https://img.shields.io/issuestats/i/github/sourceryinstitute/OpenCoarrays.svg?style=flat-square

| Avg response time              |
|:------------------------------:|
| ![Issue Stats][issue stat img] |

# Defect/Bug Report #

Today, @zbeekman  and I merged the master branch of the GCC git mirror of into the teams branch of the Sourcery Institute fork of GCC.  I just used that branch to build MPICH 3.2 and  OpenCoarrays 1.9.1 and observed the following test failures:
```
The following tests FAILED:
	 11 - async_comp_alloc_2 (Failed)
	 12 - comp_allocated_1 (Failed)
	 13 - comp_allocated_2 (Failed)
	 14 - get_array (Failed)
	 15 - get_self (Failed)
	 16 - send_array (Failed)
	 17 - get_with_offset_1d (Failed)
	 18 - whole_get_array (Failed)
	 19 - strided_get (Failed)
	 20 - strided_sendget (Failed)
	 31 - co_reduce_string (Failed)
	 37 - coarray_burgers_pde (Failed)
	 41 - convert-before-put (Failed)
	 42 - put-alloc-comp (Failed)
	 43 - get-put-alloc-comp (Failed)
Errors while running CTest
```

- OpenCoarrays Version: 1.9.1
- Fortran Compiler: gcc 8.0.0
- C compiler used for building lib: gfortran 8.0.0
- Installation method: `./install.sh -M /opt/mpich/3.2/gnu/8.0.0/ --install-prefix /opt/opencoarrays/1.9.1/gnu/8.0.0 -y -z`
- Output of `uname -a`: `Linux Sourcery-Linux-VM 4.4.0-77-generic #98-Ubuntu SMP Wed Apr 26 08:34:02 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`

- MPI library being used: 3.2
- Machine architecture and number of physical cores: 2.7 GHz Intel Core i7
- Version of CMake: 3.7.0

## Observed Behavior ##

See above.

[links]:#
[Google]: https://www.google.com
[Stack Overflow]: https://stackoverflow.com
[fenced codeblocks]: https://help.github.com/articles/creating-and-highlighting-code-blocks/
[GFortran bug tracker]: https://gcc.gnu.org/bugzilla/
[Intel MPI]: https://software.intel.com/en-us/intel-mpi-library
[MPT]: http://www.sgi.com/products/software/sps.html
[MVAPICH]: http://mvapich.cse.ohio-state.edu
[MPICH]: https://www.mpich.org
[Open-MPI]: https://www.open-mpi.org
[CONTRIBUTING.md]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md
[Helping Out]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md#helping-out
[official release]: https://github.com/sourceryinstitute/OpenCoarrays/releases
[CMake install]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/INSTALL.md#cmake-scripts
[homebrew]: https://brew.sh
[issues page]: https://github.com/sourceryinstitute/OpenCoarrays/issues
[mailing list]: https://groups.google.com/forum/#!forum/opencoarrays
[install.sh]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/install.sh

**Comment by vehre** [comment]
Really? Gfortran 8 changed the array descriptor. As long as you don't adapt, gfortran 8 will not function properly. That's the downside of carrying your own declaration of the array descriptor.

**Comment by zbeekman** [comment]
@vehre:

1. is it just lib-caf-gfortran-descriptor.h that needs to change?
1. Could we use GFortran's copy of the array descriptor?
1. Is it included in an installed header file somewhere?

A quick search through my install tree is having trouble finding any gfortran specific header files containing array descriptors

**Comment by rouson** [comment]
Applying the attached patch that @zbeekman provided reduces the test failures to the following with `OPENCOARRAYS_DEVELOPER=ON`:
```
The following tests FAILED:
	 12 - comp_allocated_1 (Failed)
	 13 - comp_allocated_2 (Failed)
	 41 - convert-before-put (Failed)
	 42 - put-alloc-comp (Failed)
	 43 - get-put-alloc-comp (Failed)
```
I think these are all the expected failures.  Correct?  I'll submit a pull request later this evening that applies the patch and closes this issue.
[patch-array-descriptor.txt](https://github.com/sourceryinstitute/OpenCoarrays/files/1301463/patch-array-descriptor.txt)

**Comment by zbeekman** [comment]
@rouson: Would you be able to `unset OPENCOARRAYS_DEVELOPER` and then
delete the CMake cache (or the whole build dir), reconfigure, rebuild and
retest? That will tell you definitely the patch resolved everything

**Comment by rouson** [comment]
@zbeekman Building with GCC 8 after `export OPENCOARRAYS_DEVELOPER=ON` gives
```
The following tests FAILED:
	 12 - comp_allocated_1 (Failed)
	 13 - comp_allocated_2 (Failed)
```
By contrast, the above two tests pass when I build with GCC 7.2.0.  Hmm... we might need some input from @vehre to resolve this

**Comment by rouson** [comment]
@vehre Any updates on this or any chance of looking at it this week?  In the interim, @zbeekman will publish a release with the updated descriptor.  We don't yet know the cause of the two new failures mentioned in my most recent comment: comp_allocated_1 and comp_allocated_2. These work with GCC 7 but not with GCC 8.

**Comment by vehre** [comment]
Have a look at the pull request for the descriptor fix: #448

**Comment by rouson** [comment]
We're all set.  I just tested the master branch built with GCC 8 and the `comp_allocated_1` and `comp_allocated_2` tests pass.  Thanks, @vehre.

@zbeekman please generate a new release at your earliest convenience.  I'll close this issue.
